File: languages/go/entity/authenticity/signature.go
package authenticity

import (
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"
)

// Signature represents a cryptographic signature along with additional metadata.
type Signature struct {
	// Signature is the cryptographic signature.
	Signature   string
	// Alg is the algorithm used for the signature.
	Alg         string
	// Kid is the key identifier associated with the signature. (public key or key ID).
	Kid         string
	// MessageHash is the hash of the message that was signed.
	MessageHash string
	// Subject is an optional field representing the subject of the signature.
	Subject     *string
}

func NewSignatureFromProto(s *proto.Signature) Signature {
	if s == nil {
		return Signature{}
	}
	return Signature{
		Signature:   s.Signature,
		Alg:         s.Alg,
		Kid:         s.Kid,
		MessageHash: s.MessageHash,
		Subject:     s.Subject,
	}
}

// GetAlg returns the SignatureAlg based on the algorithm specified in the Alg field.
func (s *Signature) GetAlg() SignatureAlg {
	if alg, ok := SignatureAlgFromProto[s.Alg]; ok {
		return alg
	}
	return UNRECOGNIZED_SIGNATURE_ALG
}

func (s Signature) ToProto() *proto.Signature {
	return &proto.Signature{
		Signature:   s.Signature,
		Alg:         s.Alg,
		Kid:         s.Kid,
		MessageHash: s.MessageHash,
		Subject:     s.Subject,
	}
}

File: languages/js/src/entity/authenticity/signature.ts
import * as proto from "../../bridge/proto/authenticity_entities";
import { SignatureAlg } from "./signature_alg";

/**
 * Represents a cryptographic signature along with additional metadata.
 */
export class Signature {
  /**
   * Is the cryptographic signature.
   */
  signature: string;
  /**
   * Is the algorithm used for the signature.
   */
  alg: string;
  /**
   * Is the key identifier associated with the signature. (public key or key ID).
   */
  kid: string;
  /**
   * Is the hash of the message that was signed.
   */
  messageHash: string;
  /**
   * Is an optional field representing the subject of the signature.
   */
  subject?: string;

  /**
   * Constructs a Signature object with the specified parameters.
   * @param messageHash 
   * @param signature 
   * @param alg 
   * @param kid 
   * @param subject 
   */
  constructor(
    messageHash: string,
    signature: string,
    alg: string,
    kid: string,
    subject?: string
  ) {
    this.signature = signature;
    this.alg = alg;
    this.kid = kid;
    this.messageHash = messageHash;
    this.subject = subject;
  }

  static fromProto(s: proto.Signature): Signature {
    return new Signature(s.messageHash, s.signature, s.alg, s.kid, s.subject);
  }

  toProto(): proto.Signature {
    return proto.Signature.fromPartial({
      signature: this.signature,
      alg: this.alg,
      kid: this.kid,
      messageHash: this.messageHash,
      subject: this.subject
    });
  }

  /**
   * Returns the SignatureAlg based on the algorithm specified in the Alg field.
   * @returns 
   */
  getAlg(): SignatureAlg {
    return SignatureAlg.fromString(this.alg);
  }
}
